cns/restserver/ipampoolmonitor.go

```diff
@@ -0,0 +1,138 @@
+package restserver
+
+import (
+	"context"
+	"sync"
+
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/Azure/azure-container-networking/cns/logger"
+	"github.com/Azure/azure-container-networking/cns/requestcontroller"
+	nnc "github.com/Azure/azure-container-networking/nodenetworkconfig/api/v1alpha"
+)
+
+var (
+	increasePoolSize = 1
+	decreasePoolSize = -1
+	doNothing        = 0
+)
+
+type IPAMPoolMonitor interface {
+	Start() error
+	UpdatePoolLimitsTransacted(batchSize int, requestThreshold float64, releaseThreshold float64)
+}
+
+type CNSIPAMPoolMonitor struct {
+	initialized bool
+
+	cns *HTTPRestService
+	rc  requestcontroller.RequestController
+
+	batchSize        int
+	requestThreshold float64
+	releaseThreshold float64
+	minimumFreeIps   int
+	maximumFreeIps   int
+	sync.RWMutex
+}
+
+func NewCNSIPAMPoolMonitor(cnsService *HTTPRestService, requestController requestcontroller.RequestController) *CNSIPAMPoolMonitor {
+	return &CNSIPAMPoolMonitor{
+		initialized: false,
+		cns:         cnsService,
+		rc:          requestController,
+	}
+}
+
+// TODO: add looping and cancellation to this, and add to CNS MAIN
+func (pm *CNSIPAMPoolMonitor) Start() error {
+
+	if pm.initialized {
+		availableIPConfigs := pm.cns.GetAvailableIPConfigs()
```

Comment:
To be optmized later: We need a more optimized solution to find available IPs, then looping the range everytime. may be keep a counter.